src/collector/otlp/service.rs

```diff
@@ -376,3 +361,80 @@ fn process_sample(
 
     Ok(())
 }
+
+fn collect_frame_list<V>(
+    loc_start: usize,
+    loc_len: usize,
+    loc_mapping: &Vec<V>,
+    profile_location_indices: &Vec<i32>,
+) -> Result<Vec<V>, Status>
+where
+    V: Copy,
+{
+    let loc_rng = loc_start..loc_start.saturating_add(loc_len);
+
+    // Collect frame list.
+    let mut frame_list = Vec::with_capacity(loc_rng.len().min(128));
+    for loc_index in loc_rng {
+        let Some(location_table_idx) = profile_location_indices.get(loc_index as usize) else {
+            return Err(Status::invalid_argument(
+                "location_indices: index is out of bounds",
+            ));
+        };
+        let Some(frame) = loc_mapping.get(*location_table_idx as usize) else {
+            return Err(Status::invalid_argument(
+                "location_table: index is out of bounds",
+            ));
+        };
+        frame_list.push(*frame);
+    }
+
+    return Ok(frame_list);
+}
+
+#[cfg(test)]
+mod tests {
+    use itertools::Itertools;
+
+    use super::*;
+
+    #[test]
+    fn sample_frame_list() -> Result<(), Status> {
+        let loc_mapping = (0..11).collect_vec();
+        let location_indices = vec![4, 9, 6, 2, 7, 4, 4, 2, 0, 1, 2, 3, 5];
+
+        assert_eq!(
+            collect_frame_list(0, 2, &loc_mapping, &location_indices)?,
+            vec![4, 9]
+        );
+        assert_eq!(
+            collect_frame_list(1, 0, &loc_mapping, &location_indices)?,
+            Vec::<i32>::new()
+        );
+        assert_eq!(
+            collect_frame_list(0, location_indices.len(), &loc_mapping, &location_indices)?,
+            location_indices
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn sample_frame_list_err() -> Result<(), Status> {
+        let loc_mapping = (0..11).collect_vec();
+
+        let loc_indices_oob = collect_frame_list(0, 3, &loc_mapping, &vec![0i32, 1i32]);
```

Comment:
Add more tests for corner cases, e.g.
```
let loc_indices_oob = collect_frame_list(1, 2, &loc_mapping, &vec![0i32, 1i32]);
let loc_indices_oob = collect_frame_list(2, 0, &loc_mapping, &vec![0i32, 1i32]);
```

Reply:
Sure: 14198b81ef3f4b01da5a139ef16ed7b638bc3bfb